app/src/main/kotlin/com/aliucord/manager/ui/viewmodel/InstallViewModel.kt

```diff
@@ -195,6 +197,79 @@ class InstallViewModel(
+
+                                        return when (name) {
+                                            "application" -> object : NodeVisitor(nv) {
+                                                override fun attr(ns: String?, name: String?, resourceId: Int, type: Int, value: Any?) {
+                                                    when (name) {
+                                                        "icon" -> squareIconId = value as Int
+                                                        "roundIcon" -> roundIconId = value as Int
+                                                    }
+                                                }
+                                            }
+                                            else -> nv
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    )
+
+                    val arscPkg = baseApk.getContent("resources.arsc").array()
+                        .let { ArscParser(it).parse() }
+                        .takeIf { it.size == 1 }
+                        ?.firstOrNull()
+                        ?: throw Error("Invalid number of packages in arsc")
+
+                    val mipmaps = arscPkg.types.values.find { it.name == "mipmap" }
+                        ?: throw Error("No mipmap type in arsc")
+
+                    val resPrefix = (arscPkg.id shl 24) or (mipmaps.id shl 16)
+
+                    val getResPaths: (Int?) -> List<String> = { targetResId ->
+                        mipmaps.configs.flatMap { config ->
+                            config.resources
+                                .filterValues { (it.spec.id or resPrefix) == targetResId }
+                                .map { (_, res) -> (res.value as Value).raw }
+                        }
+                    }
+
+                    // TODO: support patching the adaptive icon
+                    val squareIconPaths = getResPaths(squareIconId).filter { !it.endsWith("xml") }
+                    val roundIconPaths = getResPaths(roundIconId).filter { !it.endsWith("xml") }
+
+                    getResPaths(squareIconId).filter { it.endsWith("xml") }.forEach { baseApk.delete(it) }
+                    getResPaths(roundIconId).filter { it.endsWith("xml") }.forEach { baseApk.delete(it) }
+
+                    squareIconPaths.forEach {
+                        baseApk.delete(it)
+                        baseApk.add(BytesSource(application.assets.open("ic_logo_square.png").readBytes(), it, Deflater.DEFAULT_COMPRESSION))
+                    }
+                    roundIconPaths.forEach {
+                        baseApk.delete(it)
+                        baseApk.add(BytesSource(application.assets.open("ic_logo_round.png").readBytes(), it, Deflater.DEFAULT_COMPRESSION))
+                    }
+                }
+
+                log += "\n"
```

Comment:
This should probably be
```suggestion
                log += "Done\n"
```
for consistency with all the other ones